designs/odcr.md

```diff
@@ -0,0 +1,408 @@
+9. Ensure OD capacity is not automatically drifted to new capacity when a capacity reservation expires or is canceled to reduce workload disruption
+
+## Non-Goals
+
+Below lists the non-goals for _this RFC design._ Each of these items represents natural follow-ups for the initial implementation and are features we will consider based on feature requests.
+
+1. Ensure OD instances, created as fallback, can be automatically attached to an ODCR after that fact, if ODCR has availibility later (follow up needed)
+2. Support [Capacity Blocks](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/capacity-blocks-using.html) as a capacity-type -- though capacity blocks are not supported with this design, they are a natural extension of it. We could support selection on capacity blocks through the `capacityReservationSelectorTerms`.
+3. Support [Capacity Reservation Groups](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/create-cr-group.html) -- though capacity reservation groups are not supported with this design, they are a natural extension of it. We could support an additional field `reservationGroup` in the `capacityReservationSelectorTerms`.
+
+## Capacity Reservation Selection
+
+### EC2NodeClass API
+
+- Add a new field under `spec` for `capacityReservationSelectorTerms` to `EC2NodeClass` for defining which Capacity Reservation to be used for a specific `EC2NodeClass`
+  - This will allow us to attach multiple Capacity Reservations across AZs and Instance Types to a single EC2NodeClass. This capability removes the need for `Capacity Reservation Groups` for this MVP.
+- Add a new field under `status` for the found Capacity Reservations by the `spec.capacityReservationSelectorTerms` for the `EC2NodeClass`
+
+```yaml
+apiVersion: karpenter.k8s.aws/v1
+kind: EC2NodeClass
+metadata:
+  name: example-node-class
+spec:
+  # capacityReservationSelectorTerms specify selectors which are ORed together to generate
+  # a list of filters against the EC2 DescribeCapacityReservation API
+  # ID cannot be specified with any other field within a single selector
+  # All other fields are not mutually exclusive and can be combined
+  capacityReservationSelectorTerms:
+    - # The id for the Capacity Reservation
+      # Specifying '*' for this field selects all ids
+      id: String | None
+      # The id of the AWS account that owns the Capacity Reservation
+      # If no ownerID is specified, only ODCRs owned by the current account will be used
+      # Specifying '*' for this field selects all ownerIDs
+      ownerID: String | None
+      # Tags is a map of key/value tags used to select capacity reservations
+      # Specifying '*' for a value selects all values for a given tag key.
+      tags: Map | None
+status:
+  capacityReservations:
+    - # AvailabilityZone for the Capacity Reservation
+      availabilityZone: String
+      # Available Instance Count for the Capacity Reservation
+      availableInstanceCount: Integer
```

Comment:
IMO: We should be able to use this as a source-of-truth because we'll do a poll on the capacity reservations that should be eventually consistent and then update this ad-hoc as needed when we get capacity reservation launch failures. I'd agree that we have to think about the number of writes that we could get here, but so long as we update something like once per-minute I don't think the status writes should be too expensive.

Reply:
I'm not too sure I agree. We should minimize our call volume when we can, and considering the available instance count change at any time, from multiple clusters polling from the same location, this seems a bit prone to problems. I think the only way we could do it better is by injecting a call in our Create call

Reply:
> this seems a bit prone to problems

We discuss more during the implementation, but I don't think the call volume will be an issue here.